lib/MonadicRewrite.thy

```diff
@@ -14,24 +15,100 @@ imports
   LemmaBucket
 begin
 
-definition
-  monadic_rewrite :: "bool \<Rightarrow> bool \<Rightarrow> ('a \<Rightarrow> bool)
-                       \<Rightarrow> ('a, 'b) nondet_monad \<Rightarrow> ('a, 'b) nondet_monad \<Rightarrow> bool"
-where
- "monadic_rewrite F E P f g \<equiv> \<forall>s. P s \<and> (F \<longrightarrow> \<not> snd (f s))
-           \<longrightarrow> (E \<longrightarrow> f s = g s)
-            \<and> (\<not> E \<longrightarrow> fst (g s) \<subseteq> fst (f s) \<and> (snd (g s) \<longrightarrow> snd (f s)))"
+definition monadic_rewrite ::
+  "bool \<Rightarrow> bool \<Rightarrow> ('s \<Rightarrow> bool) \<Rightarrow> ('s, 'a) nondet_monad \<Rightarrow> ('s, 'a) nondet_monad \<Rightarrow> bool"
+  where
+   "monadic_rewrite F E P f g \<equiv>
+      \<forall>s. P s \<and> (F \<longrightarrow> \<not> snd (f s))
+          \<longrightarrow> (E \<longrightarrow> f s = g s) \<and> (\<not> E \<longrightarrow> fst (g s) \<subseteq> fst (f s) \<and> (snd (g s) \<longrightarrow> snd (f s)))"
 
+lemma monadic_rewrite_inst:
+  "monadic_rewrite F E P f g \<Longrightarrow> monadic_rewrite F E P f g"
+  by assumption
+
+lemma monadic_rewrite_refl:
+  "monadic_rewrite F E \<top> f f"
+  by (simp add: monadic_rewrite_def)
+
+lemma monadic_rewrite_impossible:
+  "monadic_rewrite F E \<bottom> f g"
+  by (clarsimp simp: monadic_rewrite_def)
+
+lemma monadic_rewrite_trans:
```

Comment:
Not for this PR, but in general: since we have reflexivity and transitivity, I think we should also show symmetry (when `E` is true) and anti-symmetry (when `E` is false). Not sure we'll ever really need it, but it's good to state what the basic properties are as a relation, and it might allow some more manoeuvring in proofs.

Reply:
I'm afraid you've lost me here due to gaps in my background. What is anti-symmetry and where can I read about it?
Fine to add it once I understand.

Reply:
Anti-symmetry for a relation `<=_r` is the rule`[| x <=_r y; y <=_r x |] ==> x = y`